### PR TITLE
Add `AbstractTensor::{zip,addRow}`

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/abstract_tensor.h [hunk 1]
```diff
@@ -127,6 +127,25 @@ using AbstractId = dynamic_type::DynamicType<
...
+//
+// Besides, you can stack another AbstractTensor into the current
+// AbstractTensor. For example:
+//
+// Example 11:
+//   IterDomain *id0, *id1, *id2, *id3, *id4, *id5;
+//   AbstractTensor v0({{id0, id1}, {id2, id3}});
+//   AbstractTensor v1({id4, id5});
+//   v0.stack(v1);
+// Then v0 will be [{id0, id1, id4}, {id2, id3, id5}].
```
Comment (jacobhinkle) [inline csrc/abstract_tensor.h:148]: Is `zip` just a special case of `stack` where all the input `AbstractId`s are non-vectors? I wonder if we could just implement a general stack that handles both vector and non-vector `AbstractId`s in each position and then alias it as `zip` for ergonomics.

Comment (zasdfgbnm) [reply csrc/abstract_tensor.h:148]: Good question. I think the primary concern I have is, "handles both vector and non-vector" is not very well defined. For example, if I have two abstract tensors:
```
t0 = {id0, {id1, id2}}
t1 = {id3, id4}
```
then what should the stack of `t0` with `t1` defined? Should the result be

```
{{id0, id3}, {id1, id2, id4}}
```

or 

```
{{id0, id3}, {{id1, id2}, id4}}
```

File: csrc/abstract_tensor.h [hunk 2]
```diff
@@ -207,6 +226,17 @@ struct AbstractTensor {
   // AbstractTensor is [dim0={id0, id1}, dim1={id2, id3}], then the return value
   // will be {AbstractTensor{id0, id2}, AbstractTensor{id1, id3}}.
   std::vector<AbstractTensor> unzip() const;
+
+  // Zip multiple AbstractTensors into a single AbstractTensor. For example, if
+  // the input is {AbstractTensor{id0, id2}, AbstractTensor{id1, id3}}, then the
+  // return value will be [dim0={id0, id1}, dim1={id2, id3}].
+  static AbstractTensor zip(std::vector<AbstractTensor> tensors);
+
+  // Stack another AbstractTensor into the current AbstractTensor. For example,
```
Comment (naoyam) [inline csrc/abstract_tensor.h:235]: This definition of stack sounds like concatenation?

**Review by naoyam — APPROVED** [review]
LGTM. I only left a comment on naming.

File: csrc/abstract_tensor.h (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/abstract_tensor.h:235]: It is neither like stack nor like concat... 

Stack is like below:
```
{id0, id1}
{id2, id3}
->
{{id0, id2}, {id1, id3}}
```
so it has the same meaning as `zip`.

Concat is like below:
```
{{id0}, {id1}}
{{id2}, {id3}}
->
{{id0, id2}, {id1, id3}}
```

But what this function is doing is:
```
{{id0}, {id1}}
{id2, id3}
->
{{id0, id2}, {id1, id3}}
```

Probably I should just rename it as `addRow`.

Comment (zasdfgbnm) [reply csrc/abstract_tensor.h]: renamed

**Pull request merged**